Add ingestion of X-Ray segments via UDP

hencrice opened the pull request:
**Description:**
This is largely a port of the X-Ray segment ingestion code from the existing X-Ray Daemon codebase. The main differences are:

1. in the `poll()`, I resorted to not having a buffer pool but use a local buffer instead. Details can be found in the comments for the function.
2. switched `read()` to use a more Go idiomatic return signature
3. updated the `util` package to be more idiomatic.

Note that the receiver is still incomplete. So I did not change the README file.

References for this port:
1. https://github.com/aws/aws-xray-daemon/blob/master/cmd/tracing/daemon.go#L257
2. https://github.com/aws/aws-xray-daemon/blob/master/pkg/socketconn/udp/udp.go
3. https://github.com/aws/aws-xray-daemon/blob/master/pkg/tracesegment/tracesegment.go
4. https://github.com/aws/aws-xray-daemon/blob/master/pkg/util/util.go

**Link to tracking Issue:**
Closes #430

**Testing:**
New unit tests added.

**Documentation:**
No README update needed.

hencrice committed:
Add ingestion of X-Ray segments via UDP

This is largely a port of the X-Ray segment ingestion code from the existing X-Ray Daemon codebase. The main differences are:
1. in the `poll()`, I resorted to not having a buffer pool but use a local buffer instead. Details can be found in the comments for the function.
2. switched `read()` to use a more Go idiomatic return signature

Closes #430.

hencrice committed:
Addressed lint error and boost coverage

anuraaga commented on receiver/awsxrayreceiver/internal/socketconn/udp/udp.go:
```diff
@@ -0,0 +1,55 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awsxrayreceiver/internal/socketconn"
+)
+
+// Reference for this port:
```
I think we can leave these out (though if it's context for the PR can add them as comments on the files instead of source code).

anuraaga commented on receiver/awsxrayreceiver/internal/socketconn/udp/udp.go:
```diff
@@ -0,0 +1,55 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
In open source it's fairly conventional to copy in the copyright header when copying code (copyrights generally require keeping the including license statement when copying), e.g.,

```
// Copyright 2019, OpenTelemetry Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Copyright 2018-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with the License. A copy of the License is located at
//
//     http://aws.amazon.com/apache2.0/
//
// or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
```

There might be a formal way of literally transferring the code in a way this isn't required though, maybe you can check on it?

anuraaga commented on receiver/awsxrayreceiver/internal/util/util.go:
```diff
@@ -0,0 +1,62 @@
+	var header, body []byte
+	header = *buf
+	for i := 0; i < len(bufVal); i++ {
+		if bytes.Equal(bufVal[i:i+lenSeparator], separator) {
```
Seems like it'd be better to compare against the single byte `\n` instead of comparing slices even though it's always a single byte.

anuraaga replied:
Also I think it would let us just use `bytes.IndexByte`

anuraaga commented on receiver/awsxrayreceiver/internal/util/util.go:
```diff
@@ -0,0 +1,62 @@
+const ProtocolSeparator = "\n"
+
+// SplitHeaderBody separates header and body of buf using provided separator sep, and stores in returnByte.
+func SplitHeaderBody(log *zap.Logger, buf, sep *[]byte, returnByte *[][]byte) [][]byte {
```
separator seems like a constant why do we make it a parameter here?

anuraaga commented on receiver/awsxrayreceiver/internal/util/util.go:
```diff
@@ -0,0 +1,62 @@
+// SplitHeaderBody separates header and body of buf using provided separator sep, and stores in returnByte.
+func SplitHeaderBody(log *zap.Logger, buf, sep *[]byte, returnByte *[][]byte) [][]byte {
+	if buf == nil {
+		log.Error("buffer to split is nil")
```
How about returning err instead of just logging? Can leave the decision of whether to log or not to a higher level.

anuraaga commented on receiver/awsxrayreceiver/internal/util/util.go:
```diff
@@ -0,0 +1,62 @@
+		}
+	}
+	returnByteVal := *returnByte
+	return append(returnByteVal[:0], header, body)
```
This looks like C++, can't we just return a tuple of the two arrays? Even better we can return `Header` to make it more typesafe

anuraaga commented on receiver/awsxrayreceiver/receiver.go:
```diff
@@ -0,0 +1,206 @@
+	transport = "udp"
+)
+
+// ensure the xrayReceiver implements the TraceReceiver interface
```
Is this needed, won't `newReceiver` not compile if it didn't?

anuraaga commented on receiver/awsxrayreceiver/receiver.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+// this function references the implementation in:
+// https://github.com/aws/aws-xray-daemon/blob/master/cmd/tracing/daemon.go#L275
```
You can remove this comment, it's obvious a local buffer is better here

anuraaga commented on receiver/awsxrayreceiver/receiver.go:
```diff
@@ -0,0 +1,206 @@
+// (i.e. https://github.com/aws/aws-xray-daemon/blob/d2331c8c4538e55c237b05306a3cf2e919a41ba9/pkg/bufferpool/bufferpool.go#L28)
+// is the same size as the fallBackBuffer
+// (https://github.com/aws/aws-xray-daemon/blob/master/cmd/tracing/daemon.go#L277)
+func (x *xrayReceiver) poll() {
```
Should this be `pollUdp`?

Also, you may consider moving the poll loop into `UDP` and pass back valid segment bodies through a channel so performance of `ConsumeTraces` doesn't affect the poll loop.

anuraaga replied:
I think it could also be used to make the shutdown more explicit. IIUC, currently we rely on a socket error when calling socket.close to exit this loop but that's more implicit than it could be. I think normally you would pass a signal through a channel to have the poll loop exit, upon which it signals a waitgroup.

anuraaga commented on receiver/awsxrayreceiver/receiver.go:
```diff
@@ -0,0 +1,206 @@
+			continue
+		}
+
+		// TODO: Transform payload to consumer.ConsumeTraceData. For now
```
For this PR, maybe we could at least create pdata with name copied over or something? We'd like to avoid commented out code like `// payload := slices[1]`

anuraaga reviewed (commented):
Thanks for helping with this!

hencrice committed:
Addressed feedbacks

1. Add Amazon copyright header
2. Moved references to commit message
3. Refactored a large chunk of receiver out to a separate UDP poller

hencrice committed:
Addressed lint errors

hencrice committed:
Forgot one linter error

hencrice committed:
Fixed go vet issues

anuraaga commented on receiver/awsxrayreceiver/internal/util/util.go:
```diff
@@ -0,0 +1,72 @@
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid header %w",
+			&recvErr.ErrRecoverable{Err: err})
+	} else if !header.IsValid() {
```
I was imagining checking this at the caller, but this looks better. I guess it means we don't need to return the header (this method could be called `ParseBody` or similar. I can't think of a conceivable use for the header outside of this call to `IsValid`

hencrice replied:
I can rename it in a followup PR.

anuraaga reviewed (approved):
Thanks a lot for the cleanups!

hencrice commented:
@james-bebbington Could you take a look? This PR is mostly vendor-specific code.

hencrice committed:
Merge branch 'master' into x-ray-receiver/udp

pjanotti commented on receiver/awsxrayreceiver/internal/socketconn/socketconn.go:
```diff
@@ -0,0 +1,33 @@
+package socketconn
+
+// SocketConn is an interface for socket connection.
+type SocketConn interface {
```
I suppose this is used to help with mocking in tests, can it be defined on the tests instead?

hencrice replied:
I don't think this is possible because this interface is used in both actual code (i.e. `poller.go`) and tests. And I can't move all the content of `poller.go` to be `poller_test.go`. Or maybe I misunderstood you?

pjanotti replied:
Use the original type in the actual code and cast to the convenience type only tests. It will require a cast somewhere on the tests (you can wrap it in a single function) but it avoids a construct that is intended for tests to be on production code - up to you.

hencrice replied:
Thanks for clarifying. Sorry for my ignorance but I'm still a bit confused. Are you suggesting doing things like this:

```
type poller struct {
	udpSock    *net.UDPConn
        ......
}
```

And in tests, I can define `mockSocketConn` as follows and initialize `poller` with it:
```
type mockSocketConn struct {
......
}

p := &poller{
    udpSock: &mockSocketConn{......},
}

// In some tests, cast udpSock to the mockSocketConn

p.udpSock.(*mockSocketConn)......
```

Since `net.UDPConn` is a concrete type not an interface, I thought the `poller` initialization code above will not compile. And the only way I know to make it possible for the `udpSock` field to be able to hold instances of multiple concrete type is to define an interface, which is the `SocketConn` here. Could you clarify?

Thanks!

pjanotti replied:
My mistake: I was thinking that the code was dealing with an interface, `net.PacketConn`, not a concrete type.

pjanotti commented on receiver/awsxrayreceiver/internal/udppoller/poller.go:
```diff
@@ -0,0 +1,217 @@
+		udpSock:              sock,
+		logger:               logger,
+		maxPollerCount:       cfg.NumOfPollerToStart,
+		shutDown:             make(chan bool),
```
nit: a channel of empty structs is more idiomatic

pjanotti commented on receiver/awsxrayreceiver/internal/udppoller/poller_test.go:
```diff
@@ -0,0 +1,403 @@
+// findAvailableAddress finds an available local address+port and returns it.
+// There might be race condition on the address returned by this function if
+// there's some other code that grab the address before we can listen on it.
+func findAvailableAddress() (string, error) {
```
There are already standard helpers to do this, could you please use them?

hencrice replied:
Ha yea I did find https://pkg.go.dev/go.opentelemetry.io/collector@v0.5.1-0.20200722180048-c0b3cf61a63a/testutil?tab=doc#GetAvailableLocalAddress. Unfortunately, this receiver needs UDP not TCP.

Ideally, it would be better to contribute to the collector repo so that testutil supports udp as well, then uses it in this PR. However, our timeline is a bit tight so would it be ok if I leave this PR as it is so that following work on this receiver is unblocked while I can create a separate PR to update `testutil` with UDP functionalities? Basically do things in parallel to speed things up :)

pjanotti replied:
You just need the port and then build the address from that so GetAvailablePort should be all that you need. That said it is a nit.

hencrice replied:
`GetAvailablePort` also calls `GetAvailableLocalAddress` internally so it's probably back to the same problem. Noted that this is a nit, I will still put a backlog item on my side to contribute modification to the `testutil` for testing UDP-based plugins in the future.

Thanks!

pjanotti replied:
The helper functions are basically just to get an available port the fact that they use "tcp" when finding an available one doesn't matter in this context. Anyway, it is not a blocker.

pjanotti commented on receiver/awsxrayreceiver/internal/udppoller/poller.go:
```diff
@@ -0,0 +1,217 @@
+		logger:               logger,
+		maxPollerCount:       cfg.NumOfPollerToStart,
+		shutDown:             make(chan bool),
+		segChan:              make(chan RawSegment, 30),
```
The size of this channel may be something that you want to make configurable.

hencrice replied:
Good suggestion! I normally would. However, this PR is essentially a port of the functionalities of the existing X-Ray Daemon, so I'm trying to minimize new configurations that customers may not be familiar with. We can always add more configurations in the future.

I will pull this out as a constant at the top so it's more explicit that we are hard coding something here.

pjanotti reviewed (commented):
Hi @hencrice - some small suggestions. I also need to check with someone from CNCF/OpenTelemetry if the added copyright notices are ok.

pjanotti commented:
@open-telemetry/admins @open-telemetry/technical-committee this PR is adding copyright notices below the CNCF/OpenTelemetry one. I would like a confirmation if that is ok or not. Example: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/502/files#diff-cfed4d263740d3decbf9e18976f13188R1-R23

bogdandrutu commented on receiver/awsxrayreceiver/internal/socketconn/socketconn.go:
```diff
@@ -0,0 +1,33 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright 2018-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
@lizthegrey @ccaraman @sarahnovotny do you know if this is the right way to do this?

hencrice replied:
I would love to know as well. The only reason I put this kind of header in some files is because some of the files are largely a port of the code in the existing X-Ray daemon with some relatively big changes. The X-Ray daemon itself is licensed under Apache v2.

hencrice replied:
@bogdandrutu @lizthegrey @ccaraman @sarahnovotny

I've reached out to @alolita on AWS side regarding the use of the Amazon copyright header. And she will follow up with OT committee.

hencrice commented:
> @open-telemetry/admins @open-telemetry/technical-committee this PR is adding copyright notices below the CNCF/OpenTelemetry one. I would like a confirmation if that is ok or not. Example: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/502/files#diff-cfed4d263740d3decbf9e18976f13188R1-R23

Thanks for bringing this up. I'm curious as well since this is essentially a port of some of the existing X-Ray daemon functionalities, with some relatively big changes to the original code.

bogdandrutu reviewed (changes requested):
Blocking this PR until I get the confirmation about license. Sorry for the inconvenience.

hencrice commented on receiver/awsxrayreceiver/internal/errors/errors.go:
```diff
@@ -0,0 +1,45 @@
+
+// ErrRecoverable represents an error that can be ignored
+// so that the receiver can continue to function.
+type ErrRecoverable struct {
```
hmm just want to make sure I fully understand the intended use cases for `consumererror`. Based on the existing contrib codebase, this package is only used in exporters whereas in this PR, it's a receiver. The name of the package also seems to suggest it's meant for consumers, which I believe are either processors or exporters.

All the errors I defined here are only used internally in this receiver and never passed to any OT consumer. So should I still use `consumererror`?

pjanotti replied:
If they are only used internally it is not as important, feel free to keep them.

hencrice committed:
Addressed feedbacks

hencrice committed:
Merge branch 'master' into x-ray-receiver/udp

mtwo commented:
Assigned to @pjanotti in today's meeting

hencrice commented:
> Assigned to @pjanotti in today's meeting

Oh no! I missed the SIG. Meant to ask about the licensing header issue mentioned above.

kbrockhoff commented on receiver/awsxrayreceiver/config.go:
```diff
@@ -22,7 +22,6 @@ import (
 
 const (
 	typeStr = "aws_xray"
```
exporter uses "awsxray". These two values should be consistent.

kbrockhoff commented on receiver/awsxrayreceiver/internal/udppoller/poller.go:
```diff
@@ -0,0 +1,221 @@
+			rlen, err := p.read(bufPointer)
+			if errors.As(err, &errIrrecv) {
+				p.logger.Error("irrecoverable socket read error. Exiting poller", zap.Error(err))
+				return
```
Should a new poller be launched if it returns here?

hencrice replied:
Hey @kbrockhoff thanks for the feedback.

I don't think so in this case because if it's an irrecoverable error (i.e. non-transient `net.Error` in `read()` of udppoller), that implies the UDP socket is probably closed or broken to the degree that launching more go routines will not make more progress.

kbrockhoff replied:
The X-Ray daemon takes the approach that if is there is a socket, connection or configuration issue then it kills the entire instance. This approach will not work with the collector because there may be other receivers/exporters that are still working. The collector should not stop processing metrics or logs just because traces are failing.

Tearing down and cleaning up a non-working socket and building a new one is probably the optimum approach for the final state. I am OK with just adding a TODO for now.

To make this work smoothly, we likely need to add a monitor which watches all the receivers/exports. If multiple are failing then the monitor will flip the health status to failing.

hencrice replied:
Love the idea of a health monitor. We can even come up with an interface and force all plugins to implement:
1. health ping/status report
2. best-effort restart when the plugin is unhealthy

hencrice committed:
Made typeStr consistent with xray exporter

hencrice committed:
Added TODO

lizthegrey commented:
> > Assigned to @pjanotti in today's meeting
> 
> Oh no! I missed the SIG. Meant to ask about the licensing header issue mentioned above. @bogdandrutu is there an ETA on how to proceed with the licensing issue?

If they're both Apache Licensed there's no reason to repeat the Apache boilerplate twice. However, wrt giving Amazon special mention as a copyright holder separate from just being "the opentelemetry authors", the simplest way would be to get the deletion of the amazon specific mention and replacement with copyright (firstyear)-2020 the otel authors done by an Amazon contributor covered by the CLA amazon has on file with the CNCF/OTel.

hencrice committed:
Merge from origin master

hencrice committed:
Updated copyright headers

hencrice commented:
@pjanotti @kbrockhoff @bogdandrutu I have removed the Amazon copyright header. This PR is ok to merge now. Thanks!

pjanotti commented:
@hencrice let's wait also on a green-light from @kbrockhoff

anuraaga commented:
If possible let's get a sign-off from @alolita too.

@alolita can you confirm that Amazon is transferring the copyright for existing code from the x-ray daemon to OpenTelemetry? New code is automatically handled by the CLA but I haven't seen any notion in this thread confirming we're fine for copied code. 

As in https://github.com/open-telemetry/community/issues/305 handling copied code is not trivial - even now the Java agent takes a stance of having an employee from datadog do the actual merge when copying in code, which won't apply here as it stands. @lizthegrey makes a good point that the word OpenTelemetry authors includes Amazon so the copyright isn't lost. It makes sense to me but I'm not comfortable without explicit wording from the Amazon side.

hencrice commented:
> If possible let's get a sign-off from @alolita too.
> 
> @alolita can you confirm that Amazon is transferring the copyright for existing code from the x-ray daemon to OpenTelemetry? New code is automatically handled by the CLA but I haven't seen any notion in this thread confirming we're fine for copied code.
> 
> As in [open-telemetry/community#305](https://github.com/open-telemetry/community/issues/305) handling copied code is not trivial - even now the Java agent takes a stance of having an employee from datadog do the actual merge when copying in code, which won't apply here as it stands. @lizthegrey makes a good point that the word OpenTelemetry authors includes Amazon so the copyright isn't lost. It makes sense to me but I'm not comfortable without explicit wording from the Amazon side.

Hey @anuraaga , here's what @alolita wrote in the opentelemetry-devs Slack channel: "Hi Bingfeng - the OT guidance is to not include any Amazon copyrights in the source files. They remove it if we have any copyrights associated with Amazon. Waiting for our lawyers to give clear guidance. In the meantime, @yenlinc @nicfisch don't add a copyright but please ensure you are members. We can always push back later and add if our lawyers work it out."

This is why I removed the header for now.

hencrice committed:
Merge branch 'master' of https://github.com/hencrice/opentelemetry-collector-contrib into x-ray-receiver/udp

hencrice commented:
@kbrockhoff could you take another pass? If it's ok, could you merge this? Thanks!

hencrice commented:
> @hencrice let's wait also on a green-light from @kbrockhoff

Hey @pjanotti he approved it a few days ago
